src/boot/efi/devicetree.c

```diff
@@ -104,6 +113,90 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
+        uint32_t strings_off = be32toh(dt_header->OffDTStrings);
+        uint32_t strings_size = be32toh(dt_header->SizeDTStrings);
+        uint32_t tmp;
+
+        if (struct_off % sizeof(uint32_t) != 0 || struct_size % sizeof(uint32_t) != 0 ||
+            check_add_overflow(strings_off, strings_size, &tmp) || tmp > dt_size ||
+            check_add_overflow(struct_off, struct_size, &tmp) || tmp > strings_off)
+                return NULL;
+
+        const uint32_t *cursor = (const uint32_t *) ((uint8_t *) dt_header + struct_off);
+        const char *strings_block = (const char *) ((uint8_t *) dt_header + strings_off);
+
+        size_t size_words = struct_size / sizeof(uint32_t);
+        if (check_sub_overflow(size_words, 3, &tmp))
+                return NULL;
+
+        for (size_t i = 0; i < tmp; i++) {
+                switch (be32toh(cursor[i])) {
+                case FDT_BEGIN_NODE:
+                        if (cursor[++i] != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        size_t len = be32toh(cursor[++i]);
+                        size_t name_off = be32toh(cursor[++i]);
+                        size_t len_words = DIV_ROUND_UP(len, sizeof(uint32_t));
+                        size_t s;
+
+                        if (!check_add_overflow(name_off, strlen8("compatible"), &s) &&
+                            s < strings_size && streq8(strings_block + name_off, "compatible")) {
+                                const char *c = (const char *) &cursor[++i];
+                                if (len == 0 || i + len_words > size_words || c[len - 1] != '\0')
+                                        c = NULL;
+
+                                return c;
+                        }
+                        i += len_words;
+                        break;
+                default:
+                        return NULL;
+
+                }
+        }
+
+        return NULL;
+}
+
+EFI_STATUS devicetree_match(const void *dtb_buffer, size_t dtb_length) {
+        assert(dtb_buffer);
+        const struct fdt_header *dt_header = (const struct fdt_header *)dtb_buffer;
+
+        const void *fw_dtb = find_configuration_table(MAKE_GUID_PTR(EFI_DTB_TABLE));
+        if (!fw_dtb)
+                return EFI_UNSUPPORTED;
+
+        const char *fw_compat = devicetree_get_compatible(fw_dtb);
+        if (dtb_length < sizeof(struct fdt_header) ||
+            dtb_length < be32toh(dt_header->TotalSize) ||
+            !fw_compat)
+                return EFI_INVALID_PARAMETER;
+
+        const char *compat = devicetree_get_compatible(dtb_buffer);
+        if (!compat)
+                return EFI_INVALID_PARAMETER;
+
+        /* Only matches the first compatible string from each DT */
+        if (streq8(compat, fw_compat))
+                return EFI_SUCCESS;
+        else
+                return EFI_NOT_FOUND;
```

Comment:
sounds like an ideal candidate for the ternary operator, i.e. `return streq8(…) ? EFI_SUCCESS : EFI_NOT_FOUND`